frontends/common/constantFolding.cpp

```diff
@@ -798,10 +798,18 @@ const IR::Node *DoConstantFolding::postorder(IR::Concat *e) {
         return e;
     }
 
+    // The concatenation result would be wrong if the negative value was used for the second
+    // argument (because the big_int binary representation does not match what the P4 spec expects).
+    // Therefore, we convert the second argument to a unsigned literal with the same binary
+    // representation as the P4 mandates for the signed value and concatenate these.
+    if (rt->isSigned) {
+        rt = IR::Type_Bits::get(rt->size);
+        right = new IR::Constant(rt, right->value, 10, true /* no warning */);
+    }
```

Comment:
There's no need to create a new IR::Constant here that you're going to throw away -- just take the big_int value and fix it if it is negative:

    big_int rvalue = right->value;
    if (rvalue < 0)
        rvalue &= Util::mask(rt->size);

and then use `rvalue` later instead of `right->value`

Reply:
I do not like the idea of spreading a conversion function from 2-complement signed to 2-complement unsigned over multiple places in the codebase. It is already in the `Constant::handleOverflow` (invoked from the ctor) and it is not really possible to extract it from there because of the warnings which use the constant value in the logs.

I've changed the code to avoid the unnecessary dynamic allocation at least.

Reply:
IMO this is more obscure than just doing the mask with the `big_int` value.  boost::mulitprecision specifies that bitwise operations on negative value happen as-if the values are 2s complement (regardless of what the actual underlying representation is), so that does what we want for concatenation according to the P4 spec.